doc/bmc-config.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.33.
.TH BMC-CONFIG "1" "February 2004" "bmc-config 0.1.0" "User Commands"
.SH NAME
bmc-config \- manual page for bmc-config.  IPMI BMC Configurator version 0.1.0
.SH DESCRIPTION
IPMI BMC Configurator is used to get and set BMC configurations
.SH OPTIONS
.TP
\fB\-u\fR, \fB\-\-usage\fR
Usage message
.TP
\fB\-h\fR, \fB\-\-help\fR
Show help
.TP
\fB\-V\fR, \fB\-\-version\fR
Show version
.TP
\fB\-o\fR, \fB\-\-checkout\fR
Fetch configuration information from BMC.
.TP
\fB\-i\fR, \fB\-\-commit\fR
Update configuration information to BMC
.TP
\fB\-d\fR, \fB\-\-diff\fR
Show differences between BMC and config file or key pairs.
.TP
\fB\-f\fR FILENAME, \fB\-\-filename\fR=\fIFILENAME\fR
Use this file for BMC get/set.
.TP
\fB\-k\fR "KEY=VALUE", \fB\-\-key\-pair=\fR"KEY=VALUE"
Update configuration information to BMC.  This option can be used multiple times.
.SH "EXAMPLES"
.B # bmc-config --checkout 
.PP
Show all BMC configuration to console.
.PP
.B # bmc-config --checkout --filename=bmc-data1.conf
.PP
Store all BMC configuration to bmc-data1.conf.
.PP
.B # bmc-config --diff -f bmc-data2.conf
.PP
Show all difference between BMC and bmc-data2.conf file.
.PP
.B # bmc-config --diff --key-pair="lan_conf_misc:gratuitous_arp_interval=8"
.PP
Show difference with BMC of key 'gratuitous_arp_interval' and value '4'.
.PP
.B # bmc-config --commit -f bmc-data1.conf
.PP
Commit all configuration values from bmc-data1.conf file to BMC.
.PP
.B # bmc-config --commit --key-pair="lan_conf_misc:gratuitous_arp_interval=4"
.PP
Commit key 'gratuitous_arp_interval' of value '4' to BMC.
.PP
.B # bmc-config --commit -f bmc-data-updt.conf --key-pair="lan_conf_misc:gratuitous_arp_interval=4"
.PP
Commit all configuration values from bmc-data-updt.conf and key 'gratuitous_arp_interval' of value '4' to BMC.
.PP
.SH "REPORTING BUGS"
Report bugs to <ab@gnu.org.in>.
.SH COPYRIGHT
Copyright \(co 2003-2004 FreeIPMI Core Team.  
This program is free software; you may redistribute it under the terms of
the GNU General Public License.  This program has absolutely no warranty.
.SH "SEE ALSO"
The full documentation for
.B bmc-config
is maintained as a Texinfo manual.  If the
.B info
and
.B bmc-config
programs are properly installed at your site, the command
.IP
.B info bmc-config
.PP
should give you access to the complete manual.


NEWS
FreeIPMI 1.7.1 - XX/XX/XX
-------------------------

In 2020 the IPMI standard was frozen.  So a decision was made to try
and make the FreeIPMI 1.7.X line the last stable line in FreeIPMI.  A
number of decisions were made to remove legacy support and fix minor issues,
even at the cost of breaking backwards compatibility.

Updated Features:
-----------------

Globally
--------
o Internally allow hostnames to be up to 256 characters (previously only 64).

Libraries
---------
o libfreeipmi/sensor-read: Add error for privilege level insufficient.
o In libfreeipmi, support Intel S2600BPB OEM SEL interpretations.
o Significant refactoring of Intel OEM extensions.  Names of
  variables, macros, etc. may have changed.

Tools
-----
o In ipmi-sel, support Intel S2600BPB OEM SEL interpretations.
o In ipmi-config, support configuration of Maximum_Privilege_Cipher_Suite_Id_16.
o In ipmi-sensors, support new "altbridging" workaround for Lenovo motherboards.

Smaller Bug Fixes
-----------------

o In ipmi-config, drop support for
  Rmcpplus_Conf_Privilege:Maximum_Privilege_Cipher_Suite_Id_16.  It is
  not configurable.

o Globally, fix potential output sorting error with
  --consolidate-output.  Could alter output differently than before.

o Fix many minor output errors (i.e. typos, space alignment, lack of
  newline, etc.).  Could cause issues for those who script against
  output.

o In libfreeipmi, add missing fields for BMC Message Channel SDR
  record format.

Remove Large Amount of Legacy Support
-------------------------------------

Remove backwards compatibility, legacy output/options/fields/files, or
deprecated options.  All verified to have been gone for atleast 10
years.

For details of changes see NEWS.errata

Typo Fixes:
-----------

Fix large number of typo and minor output formattings.  They have the
potential to break against existing builds or scripts or configs that
relied on previous expected output.

See NEWS.errata for details

FreeIPMI 1.6.1 - 02/02/18
-------------------------
o Add IPv6 hostname support to FreeIPMI, all of FreeIPMI can now take
  IPv6 addresses as inputs to "host" parameters, options, or inputs.
o Support significant portions of IPMI IPv6 configuration in
  libfreeipmi.
o Add --no-session option in ipmi-raw.
o Add SDR cache options to ipmi-config.
o Legacy -f short option for --flush-cache and -Q short option for
  quiet-cache.  Backwards compatible for tools that supported it
  before.
o In ipmi-oem, support Gigabyte get-bmc-services and set-bmc-services.
o Various performance improvements:
  - Remove excessive calls to secure_memset to clear memory.
  - Remove excessive memsets and clears of data.
  - Remove unnecessary "double input checks".
  - Remove expensive input checks in libfreeipmi fiid library.
    Fallout from this may include FIID_ERR_FIELD_NOT_FOUND errors in
    different fiid functions.
  - Remove unnecessary input checks in libfreeipmi fiid library.
  - Add recent 'lookups' of fields in fiid library to internal cache.
o Various minor fixes/improvements
  - Update libfreeipmi core API to use poll() instead of select(), to
    avoid issues with applications with a high number of threads.

FreeIPMI 1.5.7 - 08/16/17
-------------------------
o In libipmimonitoring, fix several mem-leak corner cases.

FreeIPMI 1.5.6 - 07/05/17
-------------------------
o In libfreeipmi locate (used by ipmi-locate), use DMI firmware in
  sysfs if it exists.
o Minor mem-leak corner case fix in libfreeipmi.
o Minor documentation fixes.
o Minor error message clarifications.

FreeIPMI 1.5.5 - 11/02/16
-------------------------
o Fix invalid flag clear in libipmiconsole that can lead to a potential
  double close on a file descriptor.
o Support Supermicro H8SGL-F OEM sensors and events.

FreeIPMI 1.5.4 - 09/30/16
-------------------------
o Various changes/fixes in libipmiconsole
  - If user retrieves file descriptor from ipmiconsole_ctx_fd, user is
    required to close it.  ipmiconsole_ctx_destroy no longer closes
    it.  This is to avoid a potential double close which can be a
    problem for multithreaded applications.  This is a change in
    behavior, but we do not believe this will affect most applications
    since most users close the file descriptor under most scenarios
    anyways.
  - ipmiconsole_ctx_destroy() should now be called to free resources
    even if ipmiconsole_engine_teardown() has been called.  This has
    been done to create consistent behavior in the API and avoid a
    former segfault possibility.  This is a change in behavior, but we
    do not believe this will affect most applications since
    ipmiconsole_engine_teardown() is only called when an application
    is being shutdown.
  - The use of IPMICONSOLE_ENGINE_CLOSE_FD has been clarified in the
    header file.  Some of the prior text was unclear.  Behavior has
    not been changed.
o In ipmi-oem, support Intel get-bmc-services and set-bmc-services commands.
o In ipmi-oem, support Gigabyte get-nic-mode and set-nic-mode commands.
o Support Gigabyte MD90-FS0-ZB OEM SEL events.

FreeIPMI 1.5.3 - 08/01/16
-------------------------
o Update Fujitsu OEM extension product id support range.
o In ipmiconsole, support --debugfile option for more difficult
  debugging scenarios.
o In all sensor related tools, increase max number of record ids and
  sensor types that can be handled.
o Fix several invalid completion code errors/values.

FreeIPMI 1.5.2 - 05/24/16
-------------------------
o Update configuration file parsing code to handle values up to 1024
  characters in length.
o Set FD_CLOEXEC for ipmi driver device files opened within
  libfreeipmi.
o Support --read-fru, --write-fru, and --device-id option in
  bmc-device.
o Support --fru-file option in ipmi-fru.
o Various library updates in libfreeipmi to handle new features.

FreeIPMI 1.5.1 - 12/01/15
-------------------------
o Update Intel Node Manager support for 3.0 specification.
  - Update API and fiid templates in libfreeipmi.

fish/extensions/bmc-config.scm
;;; bmc-config.scm: BMC configurator
;;; authors: Balamurugan <bala.a@californiadigital.com>

;;; This program is free software; you can redistribute it and/or
;;; modify it under the terms of the GNU General Public License as
;;; published by the Free Software Foundation; either version 2, or (at
;;; your option) any later version.
;;; 
;;; This program is distributed in the hope that it will be useful, but
;;; WITHOUT ANY WARRANTY; without even the implied warranty of
;;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
;;; General Public License for more details.
;;; 
;;; You should have received a copy of the GNU General Public License
;;; along with this program; if not, write to the Free Software
;;; Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA
;;; 
;;; bmc-config.scm should be automatically loaded thru init.scm

(fi-load "bc-common.scm")
(fi-load "bc-user-section.scm")
(fi-load "bc-lan-serial-channel-section.scm")
(fi-load "bc-lan-conf-section.scm")
(fi-load "bc-lan-conf-auth-section.scm")
(fi-load "bc-lan-conf-security-keys-section.scm")
(fi-load "bc-lan-conf-misc-section.scm")
(fi-load "bc-rmcpplus-conf-privilege-section.scm")
(fi-load "bc-serial-conf-section.scm")
(fi-load "bc-sol-conf-section.scm")
(fi-load "bc-pef-conf-section.scm")
(fi-load "bc-misc-section.scm")
(fi-load "bc-section.scm")

(define user1_s '("User1" 
		  "Username" 
 		  "Enable_User" 
; 		  "Clear_Password" 
		  "Password" 
		  "LAN_Enable_IPMI_Msgs" 
		  "LAN_Enable_Link_Auth" 
		  "LAN_Enable_Restrict_To_Callback" 
		  "LAN_Privilege_Limit" 
		  "LAN_Session_Limit" 
		  "Serial_Enable_IPMI_Msgs" 
		  "Serial_Enable_Link_Auth" 
		  "Serial_Enable_Restrict_To_Callback" 
		  "Serial_Privilege_Limit" 
		  "Serial_Session_Limit"))

(define user2_s '("User2" 
		  "Username" 
 		  "Enable_User" 
; 		  "Clear_Password" 
		  "Password" 
		  "LAN_Enable_IPMI_Msgs" 
		  "LAN_Enable_Link_Auth" 
		  "LAN_Enable_Restrict_To_Callback" 
		  "LAN_Privilege_Limit" 
		  "LAN_Session_Limit" 
		  "Serial_Enable_IPMI_Msgs" 
		  "Serial_Enable_Link_Auth" 
		  "Serial_Enable_Restrict_To_Callback" 
		  "Serial_Privilege_Limit" 
		  "Serial_Session_Limit"))

(define user3_s '("User3" 
		  "Username" 
 		  "Enable_User" 
; 		  "Clear_Password" 
		  "Password" 
		  "LAN_Enable_IPMI_Msgs" 
		  "LAN_Enable_Link_Auth" 
		  "LAN_Enable_Restrict_To_Callback" 
		  "LAN_Privilege_Limit" 
		  "LAN_Session_Limit" 
		  "Serial_Enable_IPMI_Msgs" 
		  "Serial_Enable_Link_Auth" 
		  "Serial_Enable_Restrict_To_Callback" 
		  "Serial_Privilege_Limit" 
		  "Serial_Session_Limit"))

(define user4_s '("User4" 
		  "Username" 
 		  "Enable_User" 
; 		  "Clear_Password" 
		  "Password" 
		  "LAN_Enable_IPMI_Msgs" 
		  "LAN_Enable_Link_Auth" 
		  "LAN_Enable_Restrict_To_Callback" 
		  "LAN_Privilege_Limit" 
		  "LAN_Session_Limit" 
		  "Serial_Enable_IPMI_Msgs" 
		  "Serial_Enable_Link_Auth" 
		  "Serial_Enable_Restrict_To_Callback" 
		  "Serial_Privilege_Limit" 
		  "Serial_Session_Limit"))

(define lan_channel_s '("LAN_Channel" 
			"Volatile_Access_Mode" 
			"Volatile_Enable_User_Level_Auth" 
			"Volatile_Enable_Per_Message_Auth" 
			"Volatile_Enable_Pef_Alerting" 
			"Volatile_Channel_Privilege_Limit" 
			"Non_Volatile_Access_Mode" 
			"Non_Volatile_Enable_User_Level_Auth" 
			"Non_Volatile_Enable_Per_Message_Auth" 
			"Non_Volatile_Enable_Pef_Alerting" 
			"Non_Volatile_Channel_Privilege_Limit"))

(define lan_conf_s '("LAN_Conf" 
		     "IP_Address_Source" 
		     "IP_Address" 
		     "MAC_Address" 
		     "Subnet_Mask" 
		     "Default_Gateway_IP_Address" 
		     "Default_Gateway_MAC_Address" 
		     "Backup_Gateway_IP_Address" 
		     "Backup_Gateway_MAC_Address"
                     "Vlan_Id_Enable"
                     "Vlan_Id"                     
                     "Vlan_Priority"))

(define lan_conf_auth_s '("LAN_Conf_Auth" 
			  "Callback_Enable_Auth_Type_None" 
			  "Callback_Enable_Auth_Type_MD2" 
			  "Callback_Enable_Auth_Type_MD5" 
			  "Callback_Enable_Auth_Type_Straight_Password" 
			  "Callback_Enable_Auth_Type_OEM_Proprietary" 
			  "User_Enable_Auth_Type_None" 
			  "User_Enable_Auth_Type_MD2" 
			  "User_Enable_Auth_Type_MD5" 
			  "User_Enable_Auth_Type_Straight_Password" 
			  "User_Enable_Auth_Type_OEM_Proprietary" 
			  "Operator_Enable_Auth_Type_None" 
			  "Operator_Enable_Auth_Type_MD2" 
			  "Operator_Enable_Auth_Type_MD5" 
			  "Operator_Enable_Auth_Type_Straight_Password" 
			  "Operator_Enable_Auth_Type_OEM_Proprietary" 
			  "Admin_Enable_Auth_Type_None" 
			  "Admin_Enable_Auth_Type_MD2" 
			  "Admin_Enable_Auth_Type_MD5" 
			  "Admin_Enable_Auth_Type_Straight_Password" 
			  "Admin_Enable_Auth_Type_OEM_Proprietary" 
			  "OEM_Enable_Auth_Type_None" 
			  "OEM_Enable_Auth_Type_MD2" 
			  "OEM_Enable_Auth_Type_MD5" 
			  "OEM_Enable_Auth_Type_Straight_Password" 
			  "OEM_Enable_Auth_Type_OEM_Proprietary"))

(define lan_conf_security_keys_s '("LAN_Conf_Security_Keys" 
                                   "K_R" 
                                   "K_G"))

(define lan_conf_misc_s '("LAN_Conf_Misc" 
			  "Enable_Gratuitous_ARPs" 
			  "Enable_ARP_Response" 
			  "Gratuitous_ARP_Interval"))

(define rmcpplus_conf_privilege_s '("Rmcpplus_Conf_Privilege" 
				    "Maximum_Privilege_Cipher_Suite_Id_0"
				    "Maximum_Privilege_Cipher_Suite_Id_1"
				    "Maximum_Privilege_Cipher_Suite_Id_2"
				    "Maximum_Privilege_Cipher_Suite_Id_3"
				    "Maximum_Privilege_Cipher_Suite_Id_4"
				    "Maximum_Privilege_Cipher_Suite_Id_5"
				    "Maximum_Privilege_Cipher_Suite_Id_6"
				    "Maximum_Privilege_Cipher_Suite_Id_7"
				    "Maximum_Privilege_Cipher_Suite_Id_8"
				    "Maximum_Privilege_Cipher_Suite_Id_9"
				    "Maximum_Privilege_Cipher_Suite_Id_10"
				    "Maximum_Privilege_Cipher_Suite_Id_11"
				    "Maximum_Privilege_Cipher_Suite_Id_12"
				    "Maximum_Privilege_Cipher_Suite_Id_13"
				    "Maximum_Privilege_Cipher_Suite_Id_14"))
  
(define serial_channel_s '("Serial_Channel" 
			   "Volatile_Access_Mode" 
			   "Volatile_Enable_User_Level_Auth" 
			   "Volatile_Enable_Per_Message_Auth" 
			   "Volatile_Enable_Pef_Alerting" 
			   "Volatile_Channel_Privilege_Limit" 
			   "Non_Volatile_Access_Mode" 
			   "Non_Volatile_Enable_User_Level_Auth" 
			   "Non_Volatile_Enable_Per_Message_Auth" 
			   "Non_Volatile_Enable_Pef_Alerting" 
			   "Non_Volatile_Channel_Privilege_Limit"))

(define serial_conf_s '("Serial_Conf" 
			"Enable_Basic_Mode" 
			"Enable_PPP_Mode" 
			"Enable_Terminal_Mode" 
			"Connect_Mode" 
			"Page_Blackout_Interval" 
			"Call_Retry_Time" 
			"Enable_DTR_Hangup" 
			"Flow_Control" 
			"Bit_Rate"))

(define pef_conf_s '("PEF_Conf" 
		     "Enable_PEF" 
		     "Enable_PEF_Event_Messages" 
		     "Enable_PEF_Startup_Delay" 
		     "Enable_PEF_Alert_Startup_Delay" 
		     "Enable_Alert_Action" 
		     "Enable_Powerdowm_Action" 
		     "Enable_Reset_Action" 
		     "Enable_Powercycle_Action" 
		     "Enable_OEM_Action" 
		     "Enable_Diagnostic_Interrupt" 
		     "PEF_Startup_Delay" 
		     "PEF_Alert_Startup_Delay"))

(define sol_conf_s '("SOL_Conf" 
		     "Enable_SOL" 
                     "SOL_Privilege_Level"
                     "Force_SOL_Payload_Authentication"
                     "Force_SOL_Payload_Encryption"
                     "Character_Accumulate_Interval"
                     "Character_Send_Threshold"
                     "SOL_Retry_Count"
                     "SOL_Retry_Interval"
                     "Non_Volatile_Bit_Rate"
                     "Volatile_Bit_Rate"
                     "SOL_Payload_Port_Number"))

(define misc_s '("Misc" 
		 "Power_Restore_Policy"))

(define (checkout-conf)
  (checkout-section user1_s (current-output-port))
  (checkout-section user2_s (current-output-port))
  (checkout-section user3_s (current-output-port))
  (checkout-section user4_s (current-output-port))
  (checkout-section lan_channel_s (current-output-port))
  (checkout-section lan_conf_s (current-output-port))
  (checkout-section lan_conf_auth_s (current-output-port))
  (checkout-section lan_conf_security_keys_s (current-output-port))
  (checkout-section lan_conf_misc_s (current-output-port))
  (checkout-section rmcpplus_conf_privilege_s (current-output-port))
  (checkout-section serial_channel_s (current-output-port))
  (checkout-section serial_conf_s (current-output-port))
  (checkout-section pef_conf_s (current-output-port))
  (checkout-section sol_conf_s (current-output-port))
  (checkout-section misc_s (current-output-port)))

(define (checkout-conf-to-file filename)
  (if (boolean? filename)
      (checkout-conf)
      (let ((fp (open-output-file filename)))
	(checkout-section user1_s fp)
	(checkout-section user2_s fp)
	(checkout-section user3_s fp)
	(checkout-section user4_s fp)
	(checkout-section lan_channel_s fp)
	(checkout-section lan_conf_s fp)
	(checkout-section lan_conf_auth_s fp)
	(checkout-section lan_conf_security_keys_s fp)
	(checkout-section lan_conf_misc_s fp)
	(checkout-section serial_channel_s fp)
	(checkout-section serial_conf_s fp)
	(checkout-section pef_conf_s fp)
	(checkout-section sol_conf_s fp)
	(checkout-section misc_s fp)
	(close fp))))

(define (validate-conf-file fp)
  (let ((section (read-section fp)))
    (if (null? section)
	#t 
	(if (validate-commit-section section)
	    (validate-conf-file fp)
	    #f))))

(define (commit-conf-file fp)
  (let ((section (read-section fp)))
    (if (null? section)
	#t
	(if (commit-section section)
	    (commit-conf-file fp)
	    #f))))

(define (validate-key-pair-list key-pair-list)
  (if (null? key-pair-list)
      #t
      (let* ((key-string (car key-pair-list))
	     (value      (cadr key-pair-list))
	     (section-data (make-section key-string value)))
	(if (list? section-data)
	    (if (validate-commit-section section-data)
		(validate-key-pair-list (cddr key-pair-list))
		#f)
	    #f))))

(define (commit-key-pair-list key-pair-list)
  (if (null? key-pair-list)
      #t
      (let* ((key-string (car key-pair-list))
	     (value      (cadr key-pair-list))
	     (section-data (make-section key-string value)))
	(if (list? section-data)
	    (if (commit-section section-data)
		(commit-key-pair-list (cddr key-pair-list))
		#f)
	    #f))))

(define (diff-key-pair-list key-pair-list)
  (if (null? key-pair-list)
      #t
      (let* ((key-string (car key-pair-list))
	     (value      (cadr key-pair-list))
	     (section-data (make-section key-string value)))
	(if (list? section-data)
	    (if (diff-section section-data)
		(diff-key-pair-list (cddr key-pair-list))
		#f)
	    #f))))

(define (diff-conf-file fp)
  (let ((section (read-section fp)))
    (if (null? section)
	#t
	(if (diff-section section)
	    (diff-conf-file fp)
	    #f))))

(define (bmc-config-main cmd-args)
  (cond 
   ((bmc-config-get-help-option cmd-args)
    (bmc-config-display-help))
   ((bmc-config-get-usage-option cmd-args)
    (bmc-config-display-usage))
   ((bmc-config-get-version-option cmd-args)
    (bmc-config-display-version))
   (else 
    (and (fi-ipmi-open cmd-args)
	 (cond 
	  ;; checkout
	  ((bmc-config-get-checkout-option cmd-args)
	   (if (not (checkout-conf-to-file 
		     (bmc-config-get-filename-option cmd-args)))
	       (set! bmc-config-exit-status -1)))
	  ;; commit
	  ((bmc-config-get-commit-option cmd-args)
	   (let ((filename (bmc-config-get-filename-option cmd-args))
		 (key-pair-list (bmc-config-get-key-pair-option cmd-args)))
	     (begin 
	       (if (string? filename)
		   (if (file-exists? filename)
		       (let ((fp (open-input-file filename)))
			 (if (validate-conf-file fp)
			     (begin 
			       (seek fp 0 SEEK_SET)
			       (if (not (commit-conf-file fp))
				   (set! bmc-config-exit-status -1)))
			     (set! bmc-config-exit-status -1))
			 (close fp))))
	       (if (list? key-pair-list)
		   (if (validate-key-pair-list key-pair-list)
		       (if (not (commit-key-pair-list key-pair-list))
			   (set! bmc-config-exit-status -1))
		       (set! bmc-config-exit-status -1))))))
	  ;; diff
	  ((bmc-config-get-diff-option cmd-args)
	   (let ((filename (bmc-config-get-filename-option cmd-args))
		 (key-pair-list (bmc-config-get-key-pair-option cmd-args)))
	     (begin 
	       (if (string? filename)
		   (if (file-exists? filename)
		       (let ((fp (open-input-file filename)))
			 (if (validate-conf-file fp)
			     (begin 
			       (seek fp 0 SEEK_SET)
			       (if (not (diff-conf-file fp))
				   (set! bmc-config-exit-status -1)))
			     (set! bmc-config-exit-status -1))
			 (close fp))))
	       (if (list? key-pair-list)
		   (if (validate-key-pair-list key-pair-list)
		       (if (not (diff-key-pair-list key-pair-list))
			   (set! bmc-config-exit-status -1))
		       (set! bmc-config-exit-status -1)))))))
	 (fi-ipmi-close)))))

(define (bmc-config args)
  "fish bmc-config main"
  (let ((cmd-args (bmc-config-argp (append (list "bmc-config") 
					   (list->strlist args)))))
    (if (list? cmd-args)
	(bmc-config-main cmd-args))))

(fi-register-command! 
 (list "bmc-config" 
       (string-append 
	"bmc-config [--no-probing] [--driver-type=IPMIDRIVER]\n"
	"           [--driver-address=DRIVERADDR] [--driver-device=DEVICE]\n"
	"           [--hostname=IPMIHOST] [--username=USERNAME]\n"
	"           [--password=PASSWORD] [--auth-type=AUTHTYPE]\n"
	"           [--priv-level=PRIVILEGE-LEVEL] [--checkout] [--commit]\n"
	"           [--diff] [--filename=FILENAME] [--key-pair=KEY-PAIR]\n"
	"           [--help] [--usage] [--version]\n"
	"\n"
	"   Displays information about BMC.")))
